urlcdd0c1ab49_emiln_JHouseFireExplorer_tgz-pJ8-dk_brics_automaton_DatatypesJ8/src/dk/brics/automaton/RegExp.java

```diff
@@ -0,0 +1,866 @@
+	}
+
+	/** 
+	 * Constructs string from parsed regular expression. 
+	 */
+	@Override
+	public String toString() {
+		return toStringBuilder(new StringBuilder()).toString();
+	}
+
+	StringBuilder toStringBuilder(StringBuilder b) {
+		switch (kind) {
+		case REGEXP_UNION:
+			b.append("(");
+			exp1.toStringBuilder(b);
+			b.append("|");
+			exp2.toStringBuilder(b);
+			b.append(")");
+			break;
+		case REGEXP_CONCATENATION:
+			exp1.toStringBuilder(b);
+			exp2.toStringBuilder(b);
+			break;
+		case REGEXP_INTERSECTION:
+			b.append("(");
+			exp1.toStringBuilder(b);
+			b.append("&");
+			exp2.toStringBuilder(b);
+			b.append(")");
+			break;
+		case REGEXP_OPTIONAL:
+			b.append("(");
+			exp1.toStringBuilder(b);
+			b.append(")?");
+			break;
+		case REGEXP_REPEAT:
+			b.append("(");
+			exp1.toStringBuilder(b);
+			b.append(")*");
+			break;
+		case REGEXP_REPEAT_MIN:
+			b.append("(");
+			exp1.toStringBuilder(b);
+			b.append("){").append(min).append(",}");
+			break;
+		case REGEXP_REPEAT_MINMAX:
+			b.append("(");
+			exp1.toStringBuilder(b);
+			b.append("){").append(min).append(",").append(max).append("}");
+			break;
+		case REGEXP_COMPLEMENT:
+			b.append("~(");
+			exp1.toStringBuilder(b);
+			b.append(")");
+			break;
+		case REGEXP_CHAR:
+			b.append("\\").append(c);
+			break;
+		case REGEXP_CHAR_RANGE:
+			b.append("[\\").append(from).append("-\\").append(to).append("]");
+			break;
+		case REGEXP_ANYCHAR:
+			b.append(".");
+			break;
+		case REGEXP_EMPTY:
+			b.append("#");
+			break;
+		case REGEXP_STRING:
+			b.append("\"").append(s).append("\"");
+			break;
+		case REGEXP_ANYSTRING:
+			b.append("@");
+			break;
+		case REGEXP_AUTOMATON:
+			b.append("<").append(s).append(">");
+			break;
+		case REGEXP_INTERVAL:
+			String s1 = Integer.toString(min);
+			String s2 = Integer.toString(max);
+			b.append("<");
+			if (digits > 0)
+				for (int i = s1.length(); i < digits; i++)
+					b.append('0');
+			b.append(s1).append("-");
+			if (digits > 0)
+				for (int i = s2.length(); i < digits; i++)
+					b.append('0');
+			b.append(s2).append(">");
+			break;
+		}
+		return b;
+	}
+
+	/** 
+	 * Returns set of automaton identifiers that occur in this regular expression. 
+	 */
+	public Set<String> getIdentifiers() {
+		HashSet<String> set = new HashSet<String>();
+		getIdentifiers(set);
+		return set;
+	}
+
+	void getIdentifiers(Set<String> set) {
+		switch (kind) {
+		case REGEXP_UNION:
+		case REGEXP_CONCATENATION:
+		case REGEXP_INTERSECTION:
+			exp1.getIdentifiers(set);
+			exp2.getIdentifiers(set);
+			break;
+		case REGEXP_OPTIONAL:
+		case REGEXP_REPEAT:
+		case REGEXP_REPEAT_MIN:
+		case REGEXP_REPEAT_MINMAX:
+		case REGEXP_COMPLEMENT:
+			exp1.getIdentifiers(set);
+			break;
+		case REGEXP_AUTOMATON:
+			set.add(s);
+			break;
+		default:
+		}
+	}
+
+	static RegExp makeUnion(RegExp exp1, RegExp exp2) {
+		RegExp r = new RegExp();
+		r.kind = Kind.REGEXP_UNION;
+		r.exp1 = exp1;
+		r.exp2 = exp2;
+		return r;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 530:534, 574:578.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.